src/components/EventQueue.cpp

```diff
@@ -0,0 +1,158 @@
+#include "EventQueue.hpp"
+#include <RTTI/RTTI_EventQueue.hpp>
+#include <Scene/BsSceneObject.h>
+#include <exception/Throw.hpp>
+
+namespace REGoth
+{
+  using SharedEMessage = EventQueue::SharedEMessage;
+
+  EventQueue::EventQueue(const bs::HSceneObject& parent)
+      : bs::Component(parent)
+  {
+  }
+
+  void EventQueue::handleMessage(SharedEMessage message, bs::HSceneObject sender)
+  {
+    message->isFirstRun = true;
+
+    if (shouldExecuteMessageInstantly(message))
+    {
+      // Pass the message to the host
+      sendMessageToHost(message, sender);
+
+      // Flag as done
+      message->deleted = true;
+    }
+    else
+    {
+      mEventQueue.push_back(message);
+    }
+  }
+
+  void EventQueue::sendMessageToHost(SharedEMessage message, bs::HSceneObject sender)
+  {
+    onExecuteEventAction(message, sender);
+
+    message->isFirstRun = false;
+  }
+
+  void EventQueue::fixedUpdate()
+  {
+    processMessageQueue();
+  }
+
+  void EventQueue::processMessageQueue()
+  {
+    // Not using iterators here, because a message might get pushed inside a callback, which
+    // would make them invalid. This has to be done before deleting the message for this very
+    // reason.
+    for (size_t i = 0, end = mEventQueue.size(); i < end; i++)
+    {
+      auto& event = mEventQueue[i];
+      if (event->deleted)
+      {
+        event->onMessageDone(event);
+      }
+    }
+
+    // Remove deleted messages from last time
+    for (auto it = mEventQueue.begin(); it != mEventQueue.end();)
+    {
+      if ((*it)->deleted)
+      {
+        it = mEventQueue.erase(it);
+      }
+      else
+      {
+        it++;
+      }
+    }
+
+    if (mEventQueue.empty()) return;
```

Comment:
This return is not required as the following for loop would be a no-op anyway if the queue is empty.